src/hotspot/share/gc/shenandoah/shenandoahHeap.cpp

```diff
@@ -2461,7 +2461,16 @@ class ShenandoahUpdateHeapRefsTask : public WorkerTask {
               // old-gen region in the most recent collection set, or if this card holds pointers to other non-specific
               // old-gen heap regions.
               if (r->is_humongous()) {
-                r->oop_iterate_humongous(&cl);
+                // Need to examine both dirty and clean cards during mixed evac.
+                ShenandoahHeapRegion* start_region = r->humongous_start_region();
+                oop obj = cast_to_oop(start_region->bottom());
+                size_t obj_size = obj->size();
+                HeapWord* end_of_object = start_region->bottom() + obj_size;
+                HeapWord* start_of_scan = r->bottom();
+                size_t card_size_in_words = CardTable::card_size_in_words();
+                size_t words_to_scan = end_of_object - start_of_scan;
```

Comment:
With some "last minute" editing, I introduced an error here.  This code should scan to end of region or end of object, whichever comes first.  William is correct to point out that we are rescanning portions of humongous objects multiple times.